File: core/src/main/java/com/pluxity/facility/facility/dto/FacilityCreateRequest.java
package com.pluxity.facility.facility.dto;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.pluxity.facility.facility.FacilityType;
import com.pluxity.facility.facility.dto.details.BuildingDetailsDto;
import com.pluxity.facility.facility.dto.details.FacilityDetailsDto;
import com.pluxity.facility.facility.dto.details.PanoramaDetailsDto;
import com.pluxity.facility.facility.dto.details.StationDetailsDto;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Builder;

@Builder
public record FacilityCreateRequest(
    @NotNull FacilityType facilityType,
    @NotBlank String name,
    @NotBlank String code,
    String description,
    Long drawingFileId,
    Long thumbnailFileId,
    @JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.EXTERNAL_PROPERTY,
        property = "facilityType"
    )
    @JsonSubTypes({
        @JsonSubTypes.Type(value = StationDetailsDto.class, name = "STATION"),
        @JsonSubTypes.Type(value = BuildingDetailsDto.class, name = "BUILDING"),
        @JsonSubTypes.Type(value = PanoramaDetailsDto.class, name = "PANORAMA")
    })
    FacilityDetailsDto details
) {
}

File: common/src/main/kotlin/com/pluxity/facility/dto/FacilityCreateRequest.kt
package com.pluxity.facility.dto

import io.swagger.v3.oas.annotations.media.Schema
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Size

data class FacilityCreateRequest(
    @field:Schema(description = "시설 이름", example = "서울역", requiredMode = Schema.RequiredMode.REQUIRED)
    @field:NotBlank(message = "이름은 필수 입니다.")
    @field:Size(max = 50, message = "이름은 최대 50자까지 입력 가능합니다.")
    val name: String,
    @field:Schema(
        description = "시설 코드",
        example = "SEOUL_STATION",
        requiredMode = Schema.RequiredMode.REQUIRED,
    )
    @field:NotBlank(message = "코드는 필수 입니다.")
    @field:Size(max = 50, message = "코드는 최대 50자까지 입력 가능합니다.")
    val code: String,
    @field:Schema(description = "시설 설명", example = "description")
    @field:Size(max = 255, message = "시설 설명은 최대 255자까지 입력 가능합니다.")
    val description: String? = null,
    @field:Schema(description = "도면 파일 ID", example = "1")
    val drawingFileId: Long? = null,
    @field:Schema(description = "썸네일파일 ID", example = "1")
    val thumbnailFileId: Long? = null,
    @field:Schema(description = "경도", example = "127")
    val lon: Double? = null,
    @field:Schema(description = "위도", example = "37")
    val lat: Double? = null,
    @field:Schema(description = "위치 관련 부가정보", example = "[{\"height\":..}]]")
    val locationMeta: String? = null,
)
